File: backend/src/ai/narrative-tarot-prose-quality.ts
/** Phase 6E.3 — Narrative result prose quality (deterministic-future sentinels). */
export type NarrativeProseQualityFailure = {
  code: 'deterministicFuture';
  pattern: string;
};

const EN: RegExp[] = [
  /\bfuture\s+promises\b/i,
  /\bthe\s+future\s+promises\b/i,
  /\bwill\s+definitely\b/i,
  /\bwill\s+certainly\b/i,
  /\binevitably\b/i,
  /\bis\s+guaranteed\s+to\b/i,
  /\bwill\s+lead\s+to\b/i,
];

const TR: RegExp[] = [
  /\bkesinlikle\s+olacak\b/i,
  /\bkesin\s+olacak\b/i,
  /\bmutlaka\s+olacak\b/i,
  /\bkaçınılmaz\s+olarak\b/i,
  /\bgaranti(?:dir|li)?\b/i,
];

const RU: RegExp[] = [
  /\bбудущее\s+обещает\b/i,
  /\bобязательно\s+произойд/i,
  /\bнеизбежно\b/i,
  /\bгарантированно\b/i,
  /\bточно\s+произойд/i,
];

function patternsFor(language: string): RegExp[] {
  if (language === 'tr') return TR;
  if (language === 'ru') return RU;
  return EN;
}

/** Collect visible Narrative result prose for prophecy scanning. */
export function narrativeVisibleProse(result: {
  summary: string;
  synthesis: string;
  advice: string;
  closingMessage: string;
  reflectionPrompt: string | null;
  dailyFocus: string | null;
  cardReadings: { text: string }[];
  relationshipInsights: { text: string }[];
  recurringCardInsights: { text: string }[];
  recurringThemeInsights: { text: string }[];
  memoryInsights: { text: string }[];
  lifeAreas: { text: string }[];
}): string {
  const parts: string[] = [
    result.summary,
    result.synthesis,
    result.advice,
    result.closingMessage,
  ];
  if (result.reflectionPrompt) parts.push(result.reflectionPrompt);
  if (result.dailyFocus) parts.push(result.dailyFocus);
  for (const list of [
    result.cardReadings,
    result.relationshipInsights,
    result.recurringCardInsights,
    result.recurringThemeInsights,
    result.memoryInsights,
    result.lifeAreas,
  ]) {
    for (const item of list) parts.push(item.text);
  }
  return parts.join('\n');
}

/** Pure detector — returns failure or null. Does not throw for expected hits. */
export function findDeterministicFuture(
  text: string,
  language: string,
): NarrativeProseQualityFailure | null {
  for (const re of patternsFor(language)) {
    if (re.test(text)) {
      return { code: 'deterministicFuture', pattern: re.source };
    }
  }
  return null;
}

File: lib/features/tarot/narrative/result/narrative_tarot_prose_quality.dart
/// Phase 6E.3 — Narrative prose quality (deterministic-future sentinels).
library;

import 'narrative_tarot_result_error.dart';
import 'narrative_tarot_structured_result.dart';

abstract final class NarrativeTarotProseQuality {
  NarrativeTarotProseQuality._();

  static final _en = <RegExp>[
    RegExp(r'\bfuture\s+promises\b', caseSensitive: false),
    RegExp(r'\bthe\s+future\s+promises\b', caseSensitive: false),
    RegExp(r'\bwill\s+definitely\b', caseSensitive: false),
    RegExp(r'\bwill\s+certainly\b', caseSensitive: false),
    RegExp(r'\binevitably\b', caseSensitive: false),
    RegExp(r'\bis\s+guaranteed\s+to\b', caseSensitive: false),
    RegExp(r'\bwill\s+lead\s+to\b', caseSensitive: false),
  ];

  static final _tr = <RegExp>[
    RegExp(r'\bkesinlikle\s+olacak\b', caseSensitive: false),
    RegExp(r'\bkesin\s+olacak\b', caseSensitive: false),
    RegExp(r'\bmutlaka\s+olacak\b', caseSensitive: false),
    RegExp(r'\bkaçınılmaz\s+olarak\b', caseSensitive: false),
    RegExp(r'\bgaranti(?:dir|li)?\b', caseSensitive: false),
  ];

  static final _ru = <RegExp>[
    RegExp(r'\bбудущее\s+обещает\b', caseSensitive: false),
    RegExp(r'\bобязательно\s+произойд', caseSensitive: false),
    RegExp(r'\bнеизбежно\b', caseSensitive: false),
    RegExp(r'\bгарантированно\b', caseSensitive: false),
    RegExp(r'\bточно\s+произойд', caseSensitive: false),
  ];

  static String visibleProse(NarrativeTarotStructuredResult r) {
    final b = StringBuffer()
      ..writeln(r.summary)
      ..writeln(r.synthesis)
      ..writeln(r.advice)
      ..writeln(r.closingMessage);
    if (r.reflectionPrompt != null) b.writeln(r.reflectionPrompt);
    if (r.dailyFocus != null) b.writeln(r.dailyFocus);
    for (final c in r.cardReadings) {
      b.writeln(c.text);
    }
    for (final x in r.relationshipInsights) {
      b.writeln(x.text);
    }
    for (final x in r.recurringCardInsights) {
      b.writeln(x.text);
    }
    for (final x in r.recurringThemeInsights) {
      b.writeln(x.text);
    }
    for (final x in r.memoryInsights) {
      b.writeln(x.text);
    }
    for (final x in r.lifeAreas) {
      b.writeln(x.text);
    }
    return b.toString();
  }

  static List<RegExp> _patterns(String languageCode) {
    if (languageCode == 'tr') return _tr;
    if (languageCode == 'ru') return _ru;
    return _en;
  }

  /// Returns matching pattern source, or null when prose is acceptable.
  static String? firstDeterministicFutureHit(
    String text, {
    required String languageCode,
  }) {
    for (final re in _patterns(languageCode)) {
      if (re.hasMatch(text)) return re.pattern;
    }
    return null;
  }

  static void validate(NarrativeTarotStructuredResult result) {
    final hit = firstDeterministicFutureHit(
      visibleProse(result),
      languageCode: result.languageCode,
    );
    if (hit != null) {
      throw NarrativeTarotResultException(
        NarrativeTarotResultErrorKind.deterministicFuture,
        hit,
      );
    }
  }
}
